sysdump/sysdump.go

```diff
@@ -0,0 +1,815 @@
+	// It is required since we don't know beforehand how many sub-tasks will be created, as they depend on the number of Cilium/Hubble/... pods found by "main" tasks.
+	// Main tasks should defer taskWg.Done() and proceed to submitting sub-tasks.
+	tasksWg sync.WaitGroup
+}
+
+func NewSysdump(k KubernetesClient, o SysdumpOptions) Sysdump {
+	return &sysdump{
+		client:  k,
+		options: o,
+		pool:    workerpool.New(int(o.WorkerCount)),
+	}
+}
+
+func (s *sysdump) Collect() error {
+	// Grab the current timestamp and create a temporary directory to hold the files.
+	t := time.Now()
+
+	// replaceTimestamp can be used to replace the special timestamp placeholder in file and directory names.
+	replaceTimestamp := func(f string) string {
+		return strings.Replace(f, timestampPlaceholderFileName, t.Format(timeFormat), -1)
+	}
+
+	d, err := ioutil.TempDir("", "*")
+	if err != nil {
+		return fmt.Errorf("failed to create temporary directory: %w", err)
+	}
+	d = filepath.Join(d, replaceTimestamp(s.options.OutputFileName))
+	if err := os.MkdirAll(d, dirMode); err != nil {
+		return fmt.Errorf("failed to create temporary directory: %w", err)
+	}
+	s.logDebug("Using %v as a temporary directory", d)
+
+	// absoluteTempPath returns the absolute path where to store the specified filename temporarily.
+	absoluteTempPath := func(f string) string {
+		return path.Join(d, replaceTimestamp(f))
+	}
+
+	// Grab the Kubernetes nodes for the target cluster.
+	s.logTask("Collecting Kubernetes nodes")
+	n, err := s.client.ListNodes(context.Background(), metav1.ListOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to collect Kubernetes nodes: %w", err)
+	}
+	s.logDebug("Finished collecting Kubernetes nodes")
+
+	// Exit if there are no nodes, as there's nothing to do.
+	if len(n.Items) == 0 {
+		s.logTask("No nodes found in the current cluster")
+		return nil
+	}
+	// If there are many nodes and no filters are specified, issue a warning and wait for a while before proceeding so the user can cancel the process.
+	if len(n.Items) > int(s.options.LargeSysdumpThreshold) && (s.options.NodeList == DefaultNodeList && s.options.LogsLimitBytes == DefaultLogsLimitBytes && s.options.LogsSinceTime == DefaultLogsSinceTime) {
+		s.logWarn("Detected a large cluster (%d nodes)", len(n.Items))
+		s.logWarn("Consider using a node filter, a custom log size limit and/or a custom log time range to decrease the size of the sysdump")
+		s.logWarn("Waiting for %s before continuing", s.options.LargeSysdumpAbortTimeout)
+		t := time.NewTicker(s.options.LargeSysdumpAbortTimeout)
+		defer t.Stop()
+		<-t.C
+	}
+
+	// Build the list of node names in which the user is interested.
+	l, err := buildNodeNameList(n, s.options.NodeList)
+	if err != nil {
+		return fmt.Errorf("failed to build node list: %w", err)
+	}
+	nodeList := l
+	s.logDebug("Restricting bugtool and logs collection to pods in %v", nodeList)
+
+	// tasks is the list of base tasks to be run.
+	tasks := []sysdumpTask{
```

Comment:
This concept is pretty neat. one thing I'd suggest to do is instead of using in-lined anonymous functions is to define pure functions and then just collect them all here.

That would hopefully avoid having a several hundred lines long function (and also optionally allow for testing individual components)

Something like
```
tasks := []sysdumpTask{
    getVersion(s),
    collectEvents(s),
    ... etc
}
```

Reply:
I think this makes sense too — I'll see how it plays with the suggestion above.